docs: document the monitoring feature (#4710)
jina-ai/jina: docs/fundamentals/flow/monitoring-flow.md

(monitoring-flow)=
# Monitor Flow

```{caution} 
The monitoring feature is still in Beta and the API is not stable yet.
```

A Jina {ref}`Flow <flow-cookbook>` exposes several core metrics that allow you to have a deeper look
on what is happening inside it. Metrics are particularly useful for building dashboards and alerts, with them, you can, for example, monitor the overall performance 

> of your Flow, detect bottleneck, alert your team when some component of your Flow are down.

Reviewer: ```suggestion
of your Flow, detect bottlenecks, alert your team when some component of your Flow is down.
```

> Jina Flows exposes metrics in the [Prometheus format](https://prometheus.io/docs/instrumenting/exposition_formats/). This is a plain text format that is understandable by both humans and machines. These metrics are intended to be scraped by

Reviewer 2: ```suggestion
Jina Flows expose metrics in the [Prometheus format](https://prometheus.io/docs/instrumenting/exposition_formats/). This is a plain text format that is understandable by both humans and machines. These metrics are intended to be scraped by
```

Jina Flow expose metrics in the [prometheus format](https://prometheus.io/docs/instrumenting/exposition_formats/) which

> is plain text that is both understandable by human and machine. These metrics are intended to be scrap by

Reviewer: ```suggestion
is plain text that is both understandable by human and machine. These metrics are intended to be scraped by
```

[prometheus](https://prometheus.io/) which is an industry standard for monitoring. 

> We encourage you to visualize metrics with [grafana](https://grafana.com/)

Reviewer: ```suggestion
We recommend you to visualize metrics with [grafana](https://grafana.com/)
```

## Using the monitoring in a Flow

A {ref}`Flow <flow-cookbook>` is composed of several Pods namely the Gateway and the Executors (potentially a Head as well). 
see the {ref}`architecture overview <architecture-overview>` for more details. Each of these Pods are microservices which communicate 
by gRPC. They will expose their own metrics using the [prometheus client](https://prometheus.io/docs/instrumenting/clientlibs/).
It means that they are as many metrics endpoints that they are Pods in your Flow. 

> Lets gave an example to illustrate it :

Reviewer: ```suggestion
Let's give an example to illustrate it :
```

````{tab} via Python API

> This example shows how to start a Flow with monitoring enable via the Python API:

Reviewer: ```suggestion
This example shows how to start a Flow with monitoring enabled via the Python API:
```

```python
from jina import Flow

with Flow(monitoring=True, port_monitoring=9090).add(
    uses='jinahub://SimpleIndexer', port_monitoring=9091
) as f:
    f.block()
```
````

````{tab} via YAML

> This example shows how to start a Flow with monitoring enable via yaml:

Reviewer: ```suggestion
This example shows how to start a Flow with monitoring enabled via yaml:
```

In a `flow.yaml` file
```yaml
!Flow
with:
  monitoring: true
  port_monitoring: 9090
executors:
- uses: jinahub://SimpleIndexer
  port_monitoring: 9091
```

```bash
jina flow --uses flow.yaml
```
````

This Flow will create two Pods, one for the Gateway, one for the SimpleIndexer Executor, and therefore it will create two 

> metrics endpoint:

Reviewer: ```suggestion
metrics endpoints:
```

The default monitoring port is `9090`, if you want to enable the monitoring on both the Gateway and the Executors you need to specify
the `prometheus_port` for the Executors. 
````

Because each Pod in a Flow exposes its own metrics, the monitoring feature can be used independently on each Pod.

> This means that you are not forced to always monitor every Pods of your Flow. For example, you could be only interested in

Reviewer 2: ```suggestion
This means that you are not forced to always monitor every Pod of your Flow. For example, you could be only interested in
```

Grammarly should've picked this up :angry:

Author: might be because of the capitalize letter

the `prometheus_port` for the Executors. 
````

> Because each Pods in a Flow expose its own metrics the monitoring feature could be used independently on each Pod.
> It means that you are not forced to always monitor each Pods of your Flow. For example, you could be only interested in
> metrics coming from the Gateway, and therefore you only activate the monitoring on it. On the other hand you might be only
> interested in monitoring a single Executor. Note that by default the monitoring is disabled everywhere.

Reviewer 3: ```suggestion
Because each Pods in a Flow exposes its own metrics, the monitoring feature can be used independently on each Pod.
This means that you are not forced to always monitor every Pods of your Flow. For example, you could be only interested in
metrics coming from the Gateway, and therefore you only activate the monitoring on it. On the other hand, you might be only
interested in monitoring a single Executor. Note that by default the monitoring is disabled everywhere.
```

To enable the monitoring you need to pass `monitoring = True` when creating the Flow.
```python
Flow(monitoring=True).add(...)
```

````{admonition} Enabling Flow
:class: hint
Passing `monitoring = True` when creating the Flow will enable the monitoring on **all the Pods** of your Flow. 
````

> If you want to enable the monitoring only on the Gateway you should once you enabled it on the Flow level, disable it for 
> all the other Executors.

Reviewer 3: Above there is an explanation saying that it is possible to monitor only certain microservices, but the explanation how to do it is only here, after the explanation of how to enable monitoring for the entire flow. I think this ordering might be more natural: 1. Say that you can enable monitoring for the entire flow 2. show how to do that 3. say that you can also monitor only select services 4. show how to do that

```suggestion
If you want to enable the monitoring only on the Gateway, you need to first enable the feature for then entire Flow, and then disable it for the Executor which you are not interested in.
```

```

On the other hand, If you want to only enable the monitoring on a given Executor you should do:
```python
Flow().add(...).add(uses=MyExecutor, monitoring=True)
```

## Available metrics

We support different metrics in the Flow, and we will later allow users to define their own.

Because all the Pods don't have the same role, they expose different kind of metrics:

### Gateway Pods

| Metrics name                       | Metrics type                                                         | Description                                                                                                                                                                                                                                                                |
|------------------------------------|----------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `jina_receiving_request_seconds`   | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measure the time elapsed between receiving a request from the client and the sending back the response.                                                                                                                                                                    |

Reviewer: ```suggestion
| `jina_receiving_request_seconds`   | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measures the time elapsed between receiving a request from the client and the sending back the response.                                                                                                                                                                    |
```

> | `jina_sending_request_seconds`     | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measure the time elapsed between sending a downstream request to an Executor/Head and receiving the response back.                                                                                                                                                         |

Reviewer: ```suggestion
| `jina_sending_request_seconds`     | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measures the time elapsed between sending a downstream request to an Executor/Head and receiving the response back.                                                                                                                                                         |
```

(monitoring-flow)=

> # Monitor Flow

Reviewer 3: I think there is a bit of an issue in the layout of this section, apologies if i saw this before and didn't say anything then. But basically, we directly parachute into monitoring only select ports etc., when that is a slightly more advanced feature to me. The simple case would be turning on monitoring for the entire flow, and I would lead with that.

Reviewer 3: ok fair

 
A Jina {ref}`Flow <flow-cookbook>` exposes several core metrics that allow you to have a deeper look
at what is happening inside it. Metrics allow you to, for example, monitor the overall performance 
of your Flow, detect bottlenecks, or alert your team when some component of your Flow is down.

Jina Flows expose metrics in the [Prometheus format](https://prometheus.io/docs/instrumenting/exposition_formats/). This is a plain text format that is understandable by both human and machine. These metrics are intended to be scraped by
[Prometheus](https://prometheus.io/),an industry standard tool for collecting and monitoring metrics.

To visualize your metrics through a dashboard, we recommend [Grafana](https://grafana.com/)

> ## Using the monitoring in a Flow

Reviewer 3: This heading and the one above say basically the same thing. What exactly is the purpose of this sub-section vs the one above?

Reviewer 3: `Monitor Flow` and `Using the monitoring in a Flow`. Should the second one be `Monitor the entire Flow`? That would make sense imo, but then further down below we should also have `Monitor select Pods in a Flow` or something like that

This means that they are as many metrics endpoints as there are Pods in your Flow. 

Let's give an example to illustrate it :

````{tab} via Python API

This example shows how to start a Flow with monitoring enabled via the Python API:

```python
from jina import Flow

with Flow(monitoring=True, port_monitoring=9090).add(
    uses='jinahub://SimpleIndexer', port_monitoring=9091
) as f:
    f.block()
```
````

````{tab} via YAML
This example shows how to start a Flow with monitoring enabled via yaml:

In a `flow.yaml` file
```yaml
!Flow
with:
  monitoring: true
  port_monitoring: 9090
executors:
- uses: jinahub://SimpleIndexer
  port_monitoring: 9091
```

```bash
jina flow --uses flow.yaml
```
````

This Flow will create two Pods, one for the Gateway, one for the SimpleIndexer Executor, and therefore it will create two 
metrics endpoints:

* `http://localhost:9090  ` for the gateway
* `http://localhost:9091  ` for the SimpleIndexer

````{admonition} Default Monitoring port
:class: hint
The default monitoring port is `9090`, if you want to enable the monitoring on both the Gateway and the Executors you need to specifiy
the `prometheus_port` for the Executors. 
````

> Because each Pods in a Flow exposes its own metrics, the monitoring feature can be used independently on each Pod.

Reviewer 3: ```suggestion
Because each Pod in a Flow exposes its own metrics, the monitoring feature can be used independently on each Pod.
```

```

On the other hand, If you want to only enable the monitoring on a given Executor you should do:
```python
Flow().add(...).add(uses=MyExecutor, monitoring=True)
```

## Available metrics

Flows support different metrics out of the box, in addition to allowing the user to define their own custom metrics.

Because not all Pods have the same role, they expose different kind of metrics:

### Gateway Pods

| Metrics name                       | Metrics type                                                         | Description                                                                                                                                                                                                                                                                |
|------------------------------------|----------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `jina_receiving_request_seconds`   | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measures the time elapsed between receiving a request from the client and the sending back the response.                                                                                                                                                                    |

Reviewer 3: ```suggestion
| `jina_receiving_request_seconds`   | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measures the time elapsed between receiving a request from the client and sending back the response.                                                                                                                                                                    |
```

| `jina_sending_request_seconds`     | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measures the time elapsed between sending a downstream request to an Executor/Head and receiving the response back.                                                                                                                                                         |

```{seealso} 
You can find more information on the different type of metrics in Prometheus [here](https://prometheus.io/docs/concepts/metric_types/#metric-types)
```

### Head Pods

| Metrics name                       | Metrics type                                                          | Description                                                                                                     |
|------------------------------------|-----------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------|

> | `jina_receiving_request_seconds`   | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary)  | Measure the time elapsed between receiving a request from the gateway and the sending back the response.        |

Reviewer 3: ```suggestion
| `jina_receiving_request_seconds`   | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary)  | Measure the time elapsed between receiving a request from the gateway and sending back the response.        |
```

| `jina_sending_request_seconds`     | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary)  | Measure the time elapsed between sending a downstream request to an Executor and receiving the response back.   |

### Executor Pods

| Metrics name                     | Metrics type                                                         | Description                                                                                                           |
|----------------------------------|----------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------|

> | `jina_receiving_request_seconds` | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measure the time elapsed between receiving a request from the gateway(or the head) and the sending back the response. |

Reviewer 3: ```suggestion
| `jina_receiving_request_seconds` | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measure the time elapsed between receiving a request from the gateway (or the head) and sending back the response. |
```

| `jina_process_request_seconds`   | [Summary](https://prometheus.io/docs/concepts/metric_types/#summary) | Measure the time spend calling the requested method                                                                   |

> | `jina_document_processed_total`  | [Counter](https://prometheus.io/docs/concepts/metric_types/#counter) | Count the number of Document processed by an Executor                                                                 |

Reviewer 3: ```suggestion
| `jina_document_processed_total`  | [Counter](https://prometheus.io/docs/concepts/metric_types/#counter) | Counts the number of Documents processed by an Executor                                                                 |
```

We encourage you to visualize metrics with [grafana](https://grafana.com/)

## Using the monitoring in a Flow

A {ref}`Flow <flow-cookbook>` is composed of several Pods namely the Gateway and the Executors (potentially a Head as well). 
see the {ref}`architecture overview <architecture-overview>` for more details. Each of these Pods are microservices which communicate 
by gRPC. They will expose their own metrics using the [prometheus client](https://prometheus.io/docs/instrumenting/clientlibs/).
It means that they are as many metrics endpoints that they are Pods in your Flow. 

Lets gave an example to illustrate it :

````{tab} via Python API

This example shows how to start a Flow with monitoring enable via the Python API:

```python
from jina import Flow

with Flow(monitoring=True, port_monitoring=9090).add(
    uses='jinahub://SimpleIndexer', monitoring=True, port_monitoring=9091
) as f:
    f.block()
```
````

````{tab} via YAML
This example shows how to start a Flow with monitoring enable via yaml:

In a `flow.yaml` file
```yaml
!Flow
with:
  monitoring: true
  port_monitoring: 9090
executors:
- uses: jinahub://SimpleIndexer
  monitoring: true

>   port_monitoring: 9091

Reviewer 4: how does this work with replicas then?

Author: The same as with the normal port. It will be random
